### PR TITLE
MULE-10517: Add information on the EventContext about the origin of its events

### PR DESCRIPTION
**Pull request by afelisatti**
DO NOT MERGE: tests running.

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/mule/model/ReflectionEntryPointResolverTestCase.java [hunk 1]
```diff
@@ -72,9 +73,10 @@ public void testExplicitMethodMatchComplexObject() throws Exception {
   public void testMethodMatchWithArguments() throws Exception {
     ReflectionEntryPointResolver resolver = new ReflectionEntryPointResolver();
     MuleEventContext eventContext =
-        new DefaultMuleEventContext(flowConstruct, Event.builder(DefaultEventContext.create(flowConstruct, TEST_CONNECTOR))
-            .message(of(new Object[] {new Apple(), new Banana()}))
-            .build());
+        new DefaultMuleEventContext(flowConstruct,
+                                    Event.builder(DefaultEventContext.create(flowConstruct, fromSingleComponent(TEST_CONNECTOR)))
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/mule/model/ReflectionEntryPointResolverTestCase.java]: don hate me, but could you create a constant `TEST_CONNECTOR_LOCATION` where `TEST_CONNECTOR` is defined?

File: core-tests/src/test/java/org/mule/runtime/core/routing/DefaultRouterResultsHandlerTestCase.java [hunk 2]
```diff
@@ -59,7 +60,7 @@ public void setupMocks() throws Exception {
 
     when(muleContext.getConfiguration()).thenReturn(mock(MuleConfiguration.class));
 
-    context = DefaultEventContext.create(flow, "test");
+    context = DefaultEventContext.create(flow, fromSingleComponent("test"));
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/routing/DefaultRouterResultsHandlerTestCase.java]: i think `TEST_CONNECTOR`'s value is `"test"`. If so you can replace this too.

File: core/src/main/java/org/mule/runtime/core/DefaultEventContext.java [hunk 3]
```diff
@@ -39,28 +39,35 @@
    * Builds a new execution context with the given parameters.
    *
    * @param flow the flow that processes events of this context.
-   * @param connectorName the name of the connector that received the first message for this context.
+   * @param location the location of the component that received the first message for this context.
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/DefaultEventContext.java]: @dfeist is this javadoc accurate after your changes?

File: core/src/main/java/org/mule/runtime/core/source/scheduler/DefaultSchedulerMessageSource.java [hunk 4]
```diff
@@ -143,7 +144,8 @@ private void pollWith(final Message request) {
 
         @Override
         public Event process() throws Exception {
-          Event event = builder(create(flowConstruct, getPollingUniqueName())).message(request).flow(flowConstruct).build();
+          Event event = builder(create(flowConstruct, fromSingleComponent(getPollingUniqueName()))).message(request)
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/source/scheduler/DefaultSchedulerMessageSource.java]: maybe it was wrong before, but isn't `DefaultSchedulerMessageSource` a source with a `ComponentLocation`?

File: modules/dsl-api/src/main/java/org/mule/runtime/dsl/api/component/config/DefaultComponentLocation.java [hunk 5]
```diff
@@ -53,6 +56,18 @@
   private String location;
   private Lock lock = new ReentrantLock();
 
+  public static DefaultComponentLocation fromSingleComponent(String component) {
```
Comment (elrodro83) [inline modules/dsl-api/src/main/java/org/mule/runtime/dsl/api/component/config/DefaultComponentLocation.java]: add a javadoc, explaining that it should not be used if possible

File: tests/performance/src/main/java/org/mule/EventBenchmark.java [hunk 6]
```diff
@@ -43,7 +44,8 @@ public void setup() throws Exception {
     flow = createFlow(muleContext);
     muleContext.getRegistry().registerFlowConstruct(flow);
     Message.Builder messageBuilder = Message.builder().payload(PAYLOAD);
-    Event.Builder eventBuilder = Event.builder(DefaultEventContext.create(flow, TEST_CONNECTOR)).message(messageBuilder.build());
+    Event.Builder eventBuilder =
+        Event.builder(DefaultEventContext.create(flow, fromSingleComponent(CONNECTOR_NAME))).message(messageBuilder.build());
```
Comment (elrodro83) [inline tests/performance/src/main/java/org/mule/EventBenchmark.java]: you changed the used constant in this class

File: tests/performance/src/main/java/org/mule/el/SimpleExpressionBenchmark.java [hunk 7]
```diff
@@ -34,7 +35,7 @@
   @Setup
   public void setup() throws MuleException {
     muleContext = createMuleContextWithServices();
-    event = Event.builder(create(createFlow(muleContext), "")).message(of(PAYLOAD)).build();
+    event = Event.builder(create(createFlow(muleContext), fromSingleComponent("test"))).message(of(PAYLOAD)).build();
```
Comment (elrodro83) [inline tests/performance/src/main/java/org/mule/el/SimpleExpressionBenchmark.java]: you changed the used constant in this class. If it has to be `"test"`, use the constant.

File: modules/dsl-api/src/main/java/org/mule/runtime/dsl/api/component/config/DefaultComponentLocation.java [hunk 8]
```diff
@@ -100,17 +121,14 @@ public TypedComponentIdentifier getComponentIdentifier() {
    */
   public String getLocation() {
     if (location == null) {
-      lock.lock();
-      try {
+      synchronized (this) {
```
Comment (elrodro83) [inline modules/dsl-api/src/main/java/org/mule/runtime/dsl/api/component/config/DefaultComponentLocation.java]: make location `volatile`

File: core/src/main/java/org/mule/runtime/core/DefaultEventContext.java (hunk 3 above)
Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/DefaultEventContext.java]: I don't think I changed anything to do with this, did I?

File: core/src/main/java/org/mule/runtime/core/DefaultEventContext.java [hunk 9]
```diff
@@ -118,6 +132,11 @@ public String getOriginatingConnectorName() {
   }
 
   @Override
+  public String getOriginatingSourceName() {
```
Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/DefaultEventContext.java]: Why do we need OriginatingLocation and OriginatingSourceName?

File: core/src/main/java/org/mule/runtime/core/api/EventContext.java [hunk 10]
```diff
@@ -51,20 +52,30 @@
...
-   * @return the name of the connector that generated the message for the first event of this context.
+   * @return the name of the component that generated the message for the first event of this context.
    */
   String getOriginatingConnectorName();
 
   /**
+   *
+   * @return the name of the source that generated the message for the first event of this context.
+   */
+  String getOriginatingSourceName();
```
Comment (dfeist) [inline core/src/main/java/org/mule/runtime/core/api/EventContext.java]: Not sure why we need three different methods

**Pull request merged**